### PR TITLE
Feature/java17 IAs

### PR DESCRIPTION
**Pull request by wederbn**
Move first exemplary IA to Java 17 using a Spring Boot SOAP service

### CODE REVIEW

**Comment by wederbn** [comment]
TODO: Wait for [ArtifactTypes PR](https://github.com/OpenTOSCA/tosca-common-artifact-types/pull/7) before merging and adapt repository.json afterwards --> DONE

**Comment by wederbn** [comment]
TODOs:
- [x] Add documentation about how to add new artifacts to the build
- [x] Remove github.ref == 'refs/pull/18/merge' from workflow before merging (only for testing purposes)
- [x] Remove pre-build WAR from repository and let Winery load the artifacts from the workflow (@lharzenetter ?)
--> Will be done in the future, thus, create a corresponding issue
- [x] Rename build products and use namespace + artifact id to uniquely identify them?

File: .github/workflows/serviceArtifactsCI.yml [hunk 1]
```diff
@@ -0,0 +1,49 @@
...
+      matrix:
+        artifactTemplates:
+          - name: DockerEngine-Interface
+            path: artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1
+            outputName: org_opentosca_artifactTemplates_DockerEngine__InterfaceDockerEngine
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - name: Set up JDK 17
+        uses: actions/setup-java@v1
```
Comment (lharzenetter) [inline .github/workflows/serviceArtifactsCI.yml]: use v2 with ``temurin`` and enable caching

File: artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/Constants.java [hunk 2]
```diff
@@ -0,0 +1,19 @@
+package org.opentosca.artifacttemplates;
+
+public class Constants {
+
+    // namespace under which the SOAP service operates
+    protected static final String NAMESPACE_URI = "http://NodeTypes.opentosca.org";
```
Comment (lharzenetter) [inline artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/Constants.java]: Shouldn't we use `http://artifacttemplates.opentsoca.org`?

File: artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/Constants.java [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+
+    // port type to use for the SOAP service
+    protected static final String PORT_TYPE_NAME = "org_opentosca_NodeTypes_DockerEngine__InterfaceDockerEnginePort";
```
Comment (lharzenetter) [inline artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/Constants.java]: I guess we should also adapt this name here. What about `org_opentosca_artifactTemplates_dockerEngine_DockerEngineInterfacePort`

File: artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/Constants.java [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+
+    // name of the XML Schema file to use as basis for the WSDL generation
+    protected static final String XSD_NAME = "model.xsd";
```
Comment (lharzenetter) [inline artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/Constants.java]: `dockerEngineInterface.xsd`?

File: artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/DockerClientHandler.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+
+public class DockerClientHandler {
+
+    private static final Logger LOG = LoggerFactory.getLogger(DockerClientHandler.class);
+
+    protected static DefaultDockerClientConfig getConfig(String DockerEngineURL, String DockerEngineCertificate) {
+        DefaultDockerClientConfig.Builder config = DefaultDockerClientConfig
+                .createDefaultConfigBuilder()
+                .withDockerHost(DockerEngineURL)
+                .withApiVersion("1.21");
```
Comment (lharzenetter) [inline artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/DockerClientHandler.java:23]: What about updating the API version?

File: artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/DockerClientHandler.java [hunk 6]
```diff
@@ -0,0 +1,76 @@
...
+     * Check if the image is already available at the Docker host
+     *
+     * @param image the image to check availability for
+     * @return The image ID if the image is available, null otherwise
+     */
+    protected static String isImageAvailable(final String image, String dockerEngineURL, String dockerEngineCertificate) {
+        LOG.info("Searching available Images...");
+        DockerClient client = DockerClientBuilder.getInstance(getConfig(dockerEngineURL, dockerEngineCertificate)).build();
+        for (final Image availImage : client.listImagesCmd().exec()) {
+            // if there are 'none' images at the Docker Engine, e.g., from a local build, this results in null as availImage
```
Comment (lharzenetter) [inline artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/DockerClientHandler.java:43]: If there are *no* images

File: artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/SoapUtil.java [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+            while (itr.hasNext()) {
+                SOAPHeaderElement header = itr.next();
+                LOG.debug("Found header with name '{}'", header.getNodeName());
+
+                if (header.getNodeName().equals(headerName)) {
+                    return header.getFirstChild();
+                }
+            }
+        } catch (javax.xml.soap.SOAPException e) {
+            e.printStackTrace();
```
Comment (lharzenetter) [inline artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/SoapUtil.java]: This should be `LOG.error(...)`

**Review by lharzenetter — CHANGES_REQUESTED** [review]

File: artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/Constants.java (hunk 3 above)
Comment (wederbn) [reply artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/Constants.java]: Sure, it just has to match `<ServiceEndpoint>/org_opentosca_NodeTypes_DockerEngine__InterfaceDockerEnginePort</ServiceEndpoint>` in the ArtifacTemplate.tosca

File: artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/Constants.java (hunk 2 above)
Comment (wederbn) [reply artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/Constants.java]: Same as below, can be changed but has to be equal to the ArtifactTemplate

File: artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/DockerClientHandler.java (hunk 6 above)
Comment (wederbn) [reply artifacttemplates/http%3A%2F%2Fopentosca.org%2Fartifacttemplates/DockerEngine_DockerEngine-Interface-w1/source/src/main/java/org/opentosca/artifacttemplates/DockerClientHandler.java:43]: The tricky thing is, that it only fails if you have images that show none as repository and tag if you execute `docker images`. So no images would be no problem.

**Review by lharzenetter — APPROVED** [review]

**Pull request merged**